### PR TITLE
[wip][dependencies] pip-compile dependencies as a build step

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

It's currently complicated to update dependencies, as it requires a (manual) recompilation of the `requirements_compiled.txt` file.

This PR adds a separate build job that runs pip compile on all input files to generate an updated `requirements_compiled.txt`, that is than exposed as an artifact in the buildkite job.

To update a dependency in the future, the process would be:

1. Open a PR that updates the respective requirements in the input requirements file
2. Wait for the CI job to run through and output the updated compiled requirements file
3. If the compilation was successful, add the contents of the file to the PR

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by can-anyscale** [comment]
I wonder if folks can just compile this locally? Or is it because they don't have a linux machine?

**Comment by krfricke** [comment]
> I wonder if folks can just compile this locally? Or is it because they don't have a linux machine?

Yes, it's mostly to streamline the compilation process. `pip-compile` can also be affected by locally installed packages/caching, so running it on a clean environment provides a common way to get a valid constraints file.

**Comment by can-anyscale** [comment]
Got you, this workflow still seems a bit troublesome to do, but better than no being able to do this at all I guess.

I can follow up on this in the future to make it possible to run this locally, which probably requires everyone to move their local dev to cloud, or running things in docker.

File: .buildkite/build.rayci.yml [hunk 1]
```diff
@@ -34,3 +34,13 @@ steps:
       - forge
       - raypy38cu118base
     job_env: forge
+
+  - label: ":tapioca: build: pip-compile dependencies"
+    instance_type: small
+    commands:
+      - rm ./python/requirements_compiled.txt
+      - ./ci/ci.sh compile_pip_dependencies
+      - cp -f ./python/requirements_compiled.txt /artifact-mount/var/
```
Comment (can-anyscale) [inline .buildkite/build.rayci.yml:43]: why not just put this into /artifact-mount?

**Review by can-anyscale — APPROVED** [review]

File: .buildkite/build.rayci.yml [hunk 2]
```diff
@@ -34,3 +34,13 @@ steps:
       - forge
       - raypy38cu118base
     job_env: forge
+
+  - label: ":tapioca: build: pip-compile dependencies"
+    instance_type: small
+    commands:
+      - rm ./python/requirements_compiled.txt
```
Comment (can-anyscale) [inline .buildkite/build.rayci.yml:41]: is this remove neccessary?

File: .buildkite/build.rayci.yml [hunk 3]
```diff
@@ -34,3 +34,13 @@ steps:
...
+      - ./ci/ci.sh compile_pip_dependencies
+      - cp -f ./python/requirements_compiled.txt /artifact-mount/var/
+    depends_on:
```
Comment (can-anyscale) [inline .buildkite/build.rayci.yml:44]: mark it soft-fail so on the day the internet decides this won't compile, it won't block people merging their PRs

File: .buildkite/build.rayci.yml (hunk 1 above)
Comment (krfricke) [reply .buildkite/build.rayci.yml:43]: This is running with non-root privileges. It looks like we can't properly modify the access flags for the root mountpoint (at least my attempts to do that failed). So a subdirectory was the easiest way to make this work.

File: .buildkite/build.rayci.yml (hunk 2 above)
Comment (krfricke) [reply .buildkite/build.rayci.yml:41]: It's a bit of a safeguard to make sure we never expose a stale requirements file

File: .buildkite/build.rayci.yml (hunk 3 above)
Comment (krfricke) [reply .buildkite/build.rayci.yml:44]: Ok!

Comment (krfricke) [reply .buildkite/build.rayci.yml:45]: Actually @can-anyscale should we keep it as hard fail to make sure incompatible dependencies are not merged?

**Review by can-anyscale — APPROVED** [review]
If we can reuse the artifact-mount so this doesn't become a special case that would be awesome. Otherwise, this is good to go!

**Pull request merged**